### PR TITLE
xtra headers

### PR DESCRIPTION
**Pull request by flash7777**
xtra headers fresh from the start

### CODE REVIEW

**Comment by flash7777** [comment]
It’s not implemented very cleanly, as this is a forward patch from 1.0.0 to 3.x.y. Please take a look at the UAG structure; I’ve taken it over in a rudimentary way, as it’s no longer present in 3.x.y.

**Comment by flash7777** [comment]
isnt there any auto-linter in git? has trailing tab(s)...   
and there's no testcase for this implementation. sorry.

**Comment by sreimers** [comment]
Not sure if this is needed, you can maybe use `usage: /uaaddheader <key>=<value> <ua-idx>`

```bash
baresip -e "/uaaddheader x-custom=test 0" -e "/uaaddheader x-custom2=test 0"
```

**Comment by flash7777** [comment]
works on register too?

File: include/baresip.h [hunk 1]
```diff
@@ -294,6 +294,12 @@ int custom_hdrs_add(struct list *hdrs, const char *name, const char *fmt, ...);
 int custom_hdrs_apply(const struct list *hdrs, custom_hdrs_h *h, void *arg);
 
 
+void debug_list_hdrs( struct list *custom_headers );
```
Comment (alfredh) [inline include/baresip.h:297]: please add a prefix to the function name ...

if the parameter `custom_headers` is not modified, it can be `const`

File: modules/contact/contact.c [hunk 2]
```diff
@@ -94,6 +94,7 @@ static int cmd_dial_contact(struct re_printf *pf, void *arg)
 
 	uri = contact_uri(cnt);
 
+
```
Comment (alfredh) [inline modules/contact/contact.c:97]: unrelated change ..

File: src/main.c [hunk 3]
```diff
@@ -75,6 +75,7 @@ static void usage(void)
 			 "\t-n <net_if>      Specify network interface\n"
 			 "\t-u <parameters>  Extra UA parameters\n"
 			 "\t-v               Verbose debug\n"
+			 "\t-x <key>:<val>   extra header\n"
```
Comment (alfredh) [inline src/main.c:78]: Please use "Extra" like the other lines. Also please specify "SIP" header.

File: src/main.c [hunk 4]
```diff
@@ -200,6 +210,36 @@ int main(int argc, char *argv[])
 			ua_eprm = optarg;
 			break;
 
+		case 'x':
+			l_ch  = strlen(optarg);
```
Comment (alfredh) [inline src/main.c:214]: here you can also use `str_len()` which is portable ...

File: src/ua.c [hunk 5]
```diff
@@ -37,6 +37,12 @@ struct ua_xhdr_filter {
 	char *hdr_name;
 };
 
+static struct {
+	struct list custom_hdrs;
+} uag = {
+	LIST_INIT,
+};
```
Comment (alfredh) [inline src/ua.c:44]: please move this block to `uag.c`

**Comment by cspiel1** [comment]
Solves issue: https://github.com/baresip/baresip/issues/3199

TODOs
- please squash the commits and set a meaningful commit message
- solve lint checks
- please add a test case

**Comment by alfredh** [comment]
I agree with the comment from @sreimers about the command `/uaaddheader`

what do you think @cspiel1

**Comment by flash7777** [comment]
My time budget is a bit tight at the moment, so I wouldn't be able to get around to implementing the changes anytime soon anyway. What's not necessary doesn't need to be included. I must have been quite naïve to add something like that without consulting the detailed documentation or studying the code, especially since it was for version 1.0.0 and I only patched it forward. Software is only as good as its documentation .... and even more so, its users. ;)

**Comment by cspiel1** [comment]
> I agree with the comment from @sreimers about the command `/uaaddheader`
> 
> what do you think @cspiel1

I realized the comment of @sreimers. But then I made a mistake and thought that the `-e` does not work for SIP REGISTER. Now I checked again that the modules like account are loaded first, this creates the `ua` objects, then the "Execute commands" from command line options are executed. Thus, the solution @sreimers mentioned should work. (I didn't test it.)

So the only benefit might be that it is a little bit more comfortable. I'd agree with you if you think this is to less benefit and we could close this.

**Comment by alfredh** [comment]
What is the next step here -- is it good enough to use the command solution ?

**Comment by alfredh** [comment]
Lets try to use the existing commands.

if you have time later and want to add your patch, please create a new PR later ...
